pkg/reconciler/pipelinerun/resources/validate_params.go

```diff
@@ -90,108 +88,58 @@ func ValidateObjectParamRequiredKeys(pipelineParameters []v1beta1.ParamSpec, pip
 }
 
 // ValidateParamArrayIndex validate if the array indexing param reference target is existent
-func ValidateParamArrayIndex(ctx context.Context, p *v1beta1.PipelineSpec, pr *v1beta1.PipelineRun) error {
+func ValidateParamArrayIndex(ctx context.Context, pipelineSpec *v1beta1.PipelineSpec, prParams []v1beta1.Param) error {
 	if !config.CheckAlphaOrBetaAPIFields(ctx) {
 		return nil
 	}
 
-	arrayParams := extractParamIndexes(p.Params, pr.Spec.Params)
+	// Collect all array params lengths
+	// Example: {"$(params.array-params)": 2, "$(params[\"array-params\"])": 2, "$(params['array-params'])": 2}
+	arrayParamsLengths := taskrun.ExtractParamArrayLengths(pipelineSpec.Params, prParams)
 
-	outofBoundParams := sets.String{}
-
-	// collect all the references
-	for i := range p.Tasks {
-		findInvalidParamArrayReferences(p.Tasks[i].Params, arrayParams, &outofBoundParams)
-		if p.Tasks[i].IsMatrixed() {
-			findInvalidParamArrayReferences(p.Tasks[i].Matrix.Params, arrayParams, &outofBoundParams)
+	paramsRefs := []string{}
```

Comment:
like $(params.ParamName), paramsRefs is used to collect all of the possible refs

Reply:
does this mean a usage of a param within a pipeline spec, or a param value that references another param? is there something else this could be called?

Reply:
> does this mean a usage of a param within a pipeline spec, or a param value that references another param

I think both, we can use the param within a spec, or pass it to a task param. 

> is there something else this could be called

Did you mean are there other places we could use this reference?

Reply:
I meant, is there any name that more clearly reflects what this variable is supposed to represent? I find "param refs" confusing because that's not really a concept in our api.

Reply:
I cannot think of any other namings. It's a reference to params. 🤔

Reply:
fair enough haha I think with some of the examples you've put in docstrings and tests it's more clear now!